quinn-proto/src/connection/mod.rs

```diff
@@ -3230,78 +3230,33 @@ impl Connection {
         now: Instant,
         packet: &mut Packet,
     ) -> Result<Option<u64>, Option<TransportError>> {
-        if !packet.header.is_protected() {
-            // Unprotected packets also don't have packet numbers
-            return Ok(None);
-        }
-        let space = packet.header.space();
-        let rx_packet = self.spaces[space].rx_packet;
-        let number = packet.header.number().ok_or(None)?.expand(rx_packet + 1);
-        let key_phase = packet.header.key_phase();
-
-        let mut crypto_update = false;
-        let crypto = if packet.header.is_0rtt() {
-            &self.zero_rtt_crypto.as_ref().unwrap().packet
-        } else if key_phase == self.key_phase || space != SpaceId::Data {
-            &self.spaces[space].crypto.as_mut().unwrap().packet.remote
-        } else if let Some(prev) = self.prev_crypto.as_ref().and_then(|crypto| {
-            // If this packet comes prior to acknowledgment of the key update by the peer,
-            if crypto.end_packet.map_or(true, |(pn, _)| number < pn) {
-                // use the previous keys.
-                Some(crypto)
-            } else {
-                // Otherwise, this must be a remotely-initiated key update, so fall through to the
-                // final case.
-                None
-            }
-        }) {
-            &prev.crypto.remote
-        } else {
-            // We're in the Data space with a key phase mismatch and either there is no locally
-            // initiated key update or the locally initiated key update was acknowledged by a
-            // lower-numbered packet. The key phase mismatch must therefore represent a new
-            // remotely-initiated key update.
-            crypto_update = true;
-            &self.next_crypto.as_ref().unwrap().remote
-        };
-
-        crypto
-            .decrypt(number, &packet.header_data, &mut packet.payload)
-            .map_err(|_| {
-                trace!("decryption failed with packet number {}", number);
-                None
-            })?;
+        let result = packet_crypto::decrypt_packet_body(
+            packet,
+            &self.spaces,
+            self.zero_rtt_crypto.as_ref(),
+            self.key_phase,
+            self.prev_crypto.as_ref(),
+            self.next_crypto.as_ref(),
+        )?;
 
-        if let Some(ref mut prev) = self.prev_crypto {
-            if prev.end_packet.is_none() && key_phase == self.key_phase {
-                // Outgoing key update newly acknowledged
-                prev.end_packet = Some((number, now));
-                self.set_key_discard_timer(now, space);
+        let pn = result.map(|d| {
```

Comment:
Style: I'd prefer this as a `match` expression with an early return, so we don't have to nest (and indent) all this code into a closure. Also not a big fan of the `d` binding here.